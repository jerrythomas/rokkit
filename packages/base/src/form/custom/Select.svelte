<script>
	import Slider from '../../list/Slider.svelte'
	import List from '../../list/List.svelte'

	import { createEventDispatcher } from 'svelte'
	const dispatch = createEventDispatcher()

	let className = ''
	export { className as class }
	export let value = null
	export let options = []
	export let fields = {}
	export let using = {}
	export let searchable = false
	let search = ''
	let visible = false

	function handleSelect(event) {
		value = event.detail
		dispatch('select', value)
	}
	$: filtered = search.trim().length
		? options.filter((opt) => opt[fields.text].includes(search))
		: options
</script>

<input-select class="flex flex-col relative {className}">
	<span
		class="flex flex-row w-full items-center"
		on:focus={() => (visible = true)}
		on:blur={() => (visible = false)}
	>
		<input type="text" bind:value={search} class:searchable class="flex" />
		<icon class="i-carbon-chevron-sort" />
	</span>
	<slot />
	{#if visible}
		<Slider>
			<List items={filtered} {fields} {using} on:select={handleSelect} />
		</Slider>
	{/if}
</input-select>
