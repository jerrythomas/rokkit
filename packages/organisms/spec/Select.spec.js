import { describe, expect, it, beforeEach, afterAll, vi } from 'vitest'
import { cleanup, render, fireEvent } from '@testing-library/svelte'
import { getPropertyValue, toHaveBeenDispatchedWith } from 'validators'
import { tick } from 'svelte'

import MockItem from './mocks/MockItem.svelte'
import Select from '../src/Select.svelte'

expect.extend({ toHaveBeenDispatchedWith })

describe('Select.svelte', () => {
	const events = ['change', 'select']
	let handlers = {}

	beforeEach(() => {
		cleanup()
		events.map((e) => (handlers[e] = vi.fn()))
	})
	afterAll(() => vi.resetAllMocks())

	it('should render string array', async () => {
		const { container } = render(Select, { options: ['a', 'b', 'c'] })
		expect(container).toBeTruthy()
		await fireEvent.click(container.querySelector('selected-item'))
		await tick()
		expect(container).toMatchSnapshot()
		let items = container.querySelectorAll('input-select list item')
		expect(items.length).toEqual(3)
		await fireEvent.click(items[1])
		await tick()
		expect(container).toMatchSnapshot()
	})

	it('should render items using field mapping', async () => {
		const { container } = render(Select, {
			options: [{ text: 1 }, { text: 2 }, { text: 3 }]
		})
		expect(container).toBeTruthy()
		await fireEvent.click(container.querySelector('selected-item'))
		await tick()
		expect(container).toMatchSnapshot()
	})

	it('should render items using custom mapping', async () => {
		const options = [{ num: 1 }, { num: 2 }, { num: 3 }]
		const { container } = render(Select, {
			options,
			value: options[1],
			fields: { text: 'num' }
		})
		expect(container).toBeTruthy()
		await fireEvent.click(container.querySelector('selected-item'))
		await tick()
		expect(container).toMatchSnapshot()
	})

	it('should render items using default component, when invalid component is provided', async () => {
		const options = [{ num: 1 }, { num: 2 }, { num: 3 }]
		const { container } = render(Select, {
			options,
			value: options[1],
			fields: { text: 'num', component: 'num' }
		})
		expect(container).toBeTruthy()
		await fireEvent.click(container.querySelector('selected-item'))
		await tick()
		expect(container).toMatchSnapshot()
	})

	it('should render with alternate class', async () => {
		const { container, component } = render(Select, {
			options: [{ num: 1 }, { num: 2 }, { num: 3 }],
			fields: { text: 'num' },
			class: 'myClass'
		})
		let classes = Array.from(container.querySelector('input-select').classList)

		expect(classes).toContain('myClass')
		component.$set({ class: 'myClass2' })
		await tick()
		classes = Array.from(container.querySelector('input-select').classList)
		expect(classes).not.toContain('myClass')
		expect(classes).toContain('myClass2')
	})

	it('should render items using custom component', async () => {
		const options = [{ num: 1, component: 'custom' }, { num: 2 }, { num: 3 }]
		const { container, component } = render(Select, {
			options,
			fields: { text: 'num' },
			using: { custom: MockItem },
			value: options[0]
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
		const item = container.querySelector('selected-item')
		await fireEvent.click(item)
		await tick()
		expect(container).toMatchSnapshot()

		component.$set({ using: { default: MockItem, custom: MockItem } })
		await tick()
		expect(container).toMatchSnapshot()
	})

	it('should emit the change event when the value changes', async () => {
		const options = [{ text: 'a' }, { text: 'b' }, { text: 'c' }]
		const { container, component } = render(Select, {
			options: options
		})

		Object.keys(handlers).map((e) => component.$on(e, handlers[e]))

		const select = container.querySelector('selected-item')
		await fireEvent.click(select)
		await tick()
		expect(container).toMatchSnapshot()

		const elements = container.querySelectorAll('input-select list item')
		await fireEvent.click(elements[1])
		await tick()
		expect(getPropertyValue(component, 'value')).toEqual(options[1])
		expect(handlers.select).toHaveBeenCalled()
		expect(handlers.change).toHaveBeenCalled()
		expect(handlers.change).toHaveBeenDispatchedWith(options[1])
		expect(handlers.select).toHaveBeenDispatchedWith(options[1])
		await tick()
		const classes = Array.from(
			container.querySelector('input-select').classList
		)
		expect(classes).not.toContain('open')
	})

	it('should open/close drop down on click and blur', async () => {
		const options = [{ text: 'a' }, { text: 'b' }, { text: 'c' }]
		const { container, component } = render(Select, {
			options: options,
			value: options[1]
		})

		Object.keys(handlers).map((e) => component.$on(e, handlers[e]))

		const wrapper = container.querySelector('input-select')
		await fireEvent.click(wrapper.querySelector('selected-item'))
		await fireEvent.focus(wrapper)

		await tick()
		let classes = Array.from(wrapper.classList)
		expect(classes).toContain('open')
		expect(wrapper).toMatchSnapshot()

		await fireEvent.blur(wrapper)
		await tick()
		classes = Array.from(wrapper.classList)
		expect(classes).not.toContain('open')
		expect(wrapper).toMatchSnapshot()
	})

	it('should close drop down on escape key', async () => {
		const options = [{ text: 'a' }, { text: 'b' }, { text: 'c' }]
		const { container, component } = render(Select, {
			options: options,
			value: options[1]
		})

		Object.keys(handlers).map((e) => component.$on(e, handlers[e]))

		const wrapper = container.querySelector('input-select')
		await fireEvent.click(wrapper.querySelector('selected-item'))

		await tick()
		let classes = Array.from(wrapper.classList)
		expect(classes).toContain('open')
		expect(wrapper).toMatchSnapshot()

		await fireEvent.keyUp(wrapper, {
			key: 'Escape'
		})
		await tick()
		classes = Array.from(wrapper.classList)
		expect(classes).not.toContain('open')
		expect(wrapper).toMatchSnapshot()
	})

	it('should handle option changes', async () => {
		const options = [{ text: 'a' }, { text: 'b' }, { text: 'c' }]
		const { container, component } = render(Select, {
			options: options,
			value: options[1]
		})

		expect(container.querySelector('selected-item')).toMatchSnapshot()
		const select = container.querySelector('selected-item')

		await fireEvent.keyDown(select, { key: 'ArrowDown' })
		await tick()
		expect(container.querySelector('scroll')).toMatchSnapshot()

		component.$set({ options: [{ text: 'a' }, { text: 'b' }] })
		await tick()
		expect(container.querySelector('scroll')).toMatchSnapshot()
	})

	it('should handle arrow keys for navigation and select', async () => {
		const options = [{ text: 'a' }, { text: 'b' }, { text: 'c' }]
		const { container, component } = render(Select, {
			options: options,
			value: options[1]
		})

		expect(container.querySelector('selected-item')).toMatchSnapshot()
		const select = container.querySelector('selected-item')

		await fireEvent.keyDown(select, { key: 'ArrowDown' })
		await tick()
		await fireEvent.keyDown(select, { key: 'ArrowDown' })
		await tick()
		await fireEvent.keyDown(select, { key: 'Enter' })
		await tick()

		expect(getPropertyValue(component, 'value')).toEqual(options[2])
		expect(container.querySelector('selected-item')).toMatchSnapshot()

		await fireEvent.keyDown(select, { key: 'ArrowUp' })
		await tick()
		await fireEvent.keyDown(select, { key: 'ArrowUp' })
		await tick()
		await fireEvent.keyDown(select, { key: 'Enter' })
		await tick()
		expect(getPropertyValue(component, 'value')).toEqual(options[1])
		expect(container.querySelector('selected-item')).toMatchSnapshot()
	})
})
