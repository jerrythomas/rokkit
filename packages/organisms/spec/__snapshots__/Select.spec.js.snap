// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select.svelte > should close drop down on escape key 1`] = `
<input-select
  class="flex flex-col relative open"
  tabindex="0"
>
  <selected-item
    aria-selected="false"
    class="w-full flex items-center"
    role="option"
    tabindex="-1"
  >
    <item>
      
       
      <p>
        b
      </p>
      
    </item>
     
    <iframe
      aria-hidden="true"
      src="about:blank"
      style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
      tabindex="-1"
    />
    <icon
      aria-label="opened"
      class="flex flex-shrink-0 items-center justify-center "
      role="button"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="selector-opened"
      />
    </icon>
  </selected-item>
   
  <scroll
    class="flex flex-col absolute w-full z-10 overflow-scroll menu"
    style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both;"
  >
    <list
      class="flex flex-col w-full flex-shrink-0 select-none list"
      role="listbox"
      tabindex="-1"
    >
       
      <item
        aria-selected="false"
        class="item"
        data-path="0"
        role="option"
      >
        
         
        <p>
          a
        </p>
        
         
      </item>
      <item
        aria-selected="true"
        class="item is-selected"
        data-path="1"
        role="option"
      >
        
         
        <p>
          b
        </p>
        
         
      </item>
      <item
        aria-selected="false"
        class="item"
        data-path="2"
        role="option"
      >
        
         
        <p>
          c
        </p>
        
         
      </item>
    </list>
  </scroll>
</input-select>
`;

exports[`Select.svelte > should close drop down on escape key 2`] = `
<input-select
  class="flex flex-col relative"
  tabindex="0"
>
  <selected-item
    aria-selected="true"
    class="w-full flex items-center"
    role="option"
    tabindex="-1"
  >
    <item>
      
       
      <p>
        b
      </p>
      
    </item>
     
    <iframe
      aria-hidden="true"
      src="about:blank"
      style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
      tabindex="-1"
    />
    <icon
      aria-label="closed"
      class="flex flex-shrink-0 items-center justify-center "
      role="button"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="selector-closed"
      />
    </icon>
  </selected-item>
   
  <scroll
    class="flex flex-col absolute w-full z-10 overflow-scroll menu"
    style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both, __svelte_2150010166_0 50ms linear 0ms 1 both;"
  >
    <list
      class="flex flex-col w-full flex-shrink-0 select-none list"
      role="listbox"
      tabindex="-1"
    >
       
      <item
        aria-selected="false"
        class="item"
        data-path="0"
        role="option"
      >
        
         
        <p>
          a
        </p>
        
         
      </item>
      <item
        aria-selected="true"
        class="item is-selected"
        data-path="1"
        role="option"
      >
        
         
        <p>
          b
        </p>
        
         
      </item>
      <item
        aria-selected="false"
        class="item"
        data-path="2"
        role="option"
      >
        
         
        <p>
          c
        </p>
        
         
      </item>
    </list>
  </scroll>
</input-select>
`;

exports[`Select.svelte > should emit the change event when the value changes 1`] = `
<body>
  <div>
    <input-select
      class="flex flex-col relative open"
      tabindex="0"
    >
      <selected-item
        aria-selected="false"
        class="w-full flex items-center"
        role="option"
        tabindex="-1"
      >
        <item>
           
          
        </item>
         
        <iframe
          aria-hidden="true"
          src="about:blank"
          style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
          tabindex="-1"
        />
        <icon
          aria-label="opened"
          class="flex flex-shrink-0 items-center justify-center "
          role="button"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-opened"
          />
        </icon>
      </selected-item>
       
      <scroll
        class="flex flex-col absolute w-full z-10 overflow-scroll menu"
        style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both;"
      >
        <list
          class="flex flex-col w-full flex-shrink-0 select-none list"
          role="listbox"
          tabindex="-1"
        >
           
          <item
            aria-selected="false"
            class="item"
            data-path="0"
            role="option"
          >
            
             
            <p>
              a
            </p>
            
             
          </item>
          <item
            aria-selected="false"
            class="item"
            data-path="1"
            role="option"
          >
            
             
            <p>
              b
            </p>
            
             
          </item>
          <item
            aria-selected="false"
            class="item"
            data-path="2"
            role="option"
          >
            
             
            <p>
              c
            </p>
            
             
          </item>
        </list>
      </scroll>
    </input-select>
  </div>
</body>
`;

exports[`Select.svelte > should handle arrow keys for navigation and select 1`] = `
<selected-item
  aria-selected="true"
  class="w-full flex items-center"
  role="option"
  tabindex="-1"
>
  <item>
    
     
    <p>
      b
    </p>
    
  </item>
   
  <icon
    aria-label="closed"
    class="flex flex-shrink-0 items-center justify-center "
    role="button"
    tabindex="-1"
  >
    <i
      aria-hidden="true"
      class="selector-closed"
    />
  </icon>
  <iframe
    aria-hidden="true"
    src="about:blank"
    style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
    tabindex="-1"
  />
</selected-item>
`;

exports[`Select.svelte > should handle arrow keys for navigation and select 2`] = `
<selected-item
  aria-selected="true"
  class="w-full flex items-center"
  role="option"
  tabindex="-1"
>
  <item>
    
     
    <p>
      c
    </p>
    
  </item>
   
  <iframe
    aria-hidden="true"
    src="about:blank"
    style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
    tabindex="-1"
  />
  <icon
    aria-label="closed"
    class="flex flex-shrink-0 items-center justify-center "
    role="button"
    tabindex="-1"
  >
    <i
      aria-hidden="true"
      class="selector-closed"
    />
  </icon>
</selected-item>
`;

exports[`Select.svelte > should handle arrow keys for navigation and select 3`] = `
<selected-item
  aria-selected="true"
  class="w-full flex items-center"
  role="option"
  tabindex="-1"
>
  <item>
    
     
    <p>
      b
    </p>
    
  </item>
   
  <iframe
    aria-hidden="true"
    src="about:blank"
    style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
    tabindex="-1"
  />
  <icon
    aria-label="closed"
    class="flex flex-shrink-0 items-center justify-center "
    role="button"
    tabindex="-1"
  >
    <i
      aria-hidden="true"
      class="selector-closed"
    />
  </icon>
</selected-item>
`;

exports[`Select.svelte > should handle option changes 1`] = `
<selected-item
  aria-selected="true"
  class="w-full flex items-center"
  role="option"
  tabindex="-1"
>
  <item>
    
     
    <p>
      b
    </p>
    
  </item>
   
  <icon
    aria-label="closed"
    class="flex flex-shrink-0 items-center justify-center "
    role="button"
    tabindex="-1"
  >
    <i
      aria-hidden="true"
      class="selector-closed"
    />
  </icon>
  <iframe
    aria-hidden="true"
    src="about:blank"
    style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
    tabindex="-1"
  />
</selected-item>
`;

exports[`Select.svelte > should handle option changes 2`] = `
<scroll
  class="flex flex-col absolute w-full z-10 overflow-scroll menu"
  style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both;"
>
  <list
    class="flex flex-col w-full flex-shrink-0 select-none list"
    role="listbox"
    tabindex="-1"
  >
     
    <item
      aria-selected="false"
      class="item"
      data-path="0"
      role="option"
    >
      
       
      <p>
        a
      </p>
      
       
    </item>
    <item
      aria-selected="true"
      class="item is-selected"
      data-path="1"
      role="option"
    >
      
       
      <p>
        b
      </p>
      
       
    </item>
    <item
      aria-selected="false"
      class="item"
      data-path="2"
      role="option"
    >
      
       
      <p>
        c
      </p>
      
       
    </item>
  </list>
</scroll>
`;

exports[`Select.svelte > should handle option changes 3`] = `
<scroll
  class="flex flex-col absolute w-full z-10 overflow-scroll menu"
  style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both;"
>
  <list
    class="flex flex-col w-full flex-shrink-0 select-none list"
    role="listbox"
    tabindex="-1"
  >
     
    <item
      aria-selected="false"
      class="item"
      data-path="0"
      role="option"
    >
      
       
      <p>
        a
      </p>
      
       
    </item>
    <item
      aria-selected="false"
      class="item"
      data-path="1"
      role="option"
    >
      
       
      <p>
        b
      </p>
      
       
    </item>
  </list>
</scroll>
`;

exports[`Select.svelte > should open/close drop down on click and blur 1`] = `
<input-select
  class="flex flex-col relative open"
  tabindex="0"
>
  <selected-item
    aria-selected="false"
    class="w-full flex items-center"
    role="option"
    tabindex="-1"
  >
    <item>
      
       
      <p>
        b
      </p>
      
    </item>
     
    <iframe
      aria-hidden="true"
      src="about:blank"
      style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
      tabindex="-1"
    />
    <icon
      aria-label="opened"
      class="flex flex-shrink-0 items-center justify-center "
      role="button"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="selector-opened"
      />
    </icon>
  </selected-item>
   
  <scroll
    class="flex flex-col absolute w-full z-10 overflow-scroll menu"
    style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both;"
  >
    <list
      class="flex flex-col w-full flex-shrink-0 select-none list"
      role="listbox"
      tabindex="-1"
    >
       
      <item
        aria-selected="false"
        class="item"
        data-path="0"
        role="option"
      >
        
         
        <p>
          a
        </p>
        
         
      </item>
      <item
        aria-selected="true"
        class="item is-selected"
        data-path="1"
        role="option"
      >
        
         
        <p>
          b
        </p>
        
         
      </item>
      <item
        aria-selected="false"
        class="item"
        data-path="2"
        role="option"
      >
        
         
        <p>
          c
        </p>
        
         
      </item>
    </list>
  </scroll>
</input-select>
`;

exports[`Select.svelte > should open/close drop down on click and blur 2`] = `
<input-select
  class="flex flex-col relative"
  tabindex="0"
>
  <selected-item
    aria-selected="true"
    class="w-full flex items-center"
    role="option"
    tabindex="-1"
  >
    <item>
      
       
      <p>
        b
      </p>
      
    </item>
     
    <iframe
      aria-hidden="true"
      src="about:blank"
      style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
      tabindex="-1"
    />
    <icon
      aria-label="closed"
      class="flex flex-shrink-0 items-center justify-center "
      role="button"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="selector-closed"
      />
    </icon>
  </selected-item>
   
  <scroll
    class="flex flex-col absolute w-full z-10 overflow-scroll menu"
    style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both, __svelte_2150010166_0 50ms linear 0ms 1 both;"
  >
    <list
      class="flex flex-col w-full flex-shrink-0 select-none list"
      role="listbox"
      tabindex="-1"
    >
       
      <item
        aria-selected="false"
        class="item"
        data-path="0"
        role="option"
      >
        
         
        <p>
          a
        </p>
        
         
      </item>
      <item
        aria-selected="true"
        class="item is-selected"
        data-path="1"
        role="option"
      >
        
         
        <p>
          b
        </p>
        
         
      </item>
      <item
        aria-selected="false"
        class="item"
        data-path="2"
        role="option"
      >
        
         
        <p>
          c
        </p>
        
         
      </item>
    </list>
  </scroll>
</input-select>
`;

exports[`Select.svelte > should render items using custom component 1`] = `
<body>
  <div>
    <input-select
      class="flex flex-col relative "
      tabindex="0"
    >
      <selected-item
        aria-selected="true"
        class="w-full flex items-center"
        role="option"
        tabindex="-1"
      >
        <item>
          
           
          <p>
            1
          </p>
          
        </item>
         
        <icon
          aria-label="closed"
          class="flex flex-shrink-0 items-center justify-center "
          role="button"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-closed"
          />
        </icon>
        <iframe
          aria-hidden="true"
          src="about:blank"
          style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
          tabindex="-1"
        />
      </selected-item>
       
    </input-select>
  </div>
</body>
`;

exports[`Select.svelte > should render items using custom component 2`] = `
<body>
  <div>
    <input-select
      class="flex flex-col relative open"
      tabindex="0"
    >
      <selected-item
        aria-selected="false"
        class="w-full flex items-center"
        role="option"
        tabindex="-1"
      >
        <item>
          
           
          <p>
            1
          </p>
          
        </item>
         
        <iframe
          aria-hidden="true"
          src="about:blank"
          style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
          tabindex="-1"
        />
        <icon
          aria-label="opened"
          class="flex flex-shrink-0 items-center justify-center "
          role="button"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-opened"
          />
        </icon>
      </selected-item>
       
      <scroll
        class="flex flex-col absolute w-full z-10 overflow-scroll menu"
        style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both;"
      >
        <list
          class="flex flex-col w-full flex-shrink-0 select-none list"
          role="listbox"
          tabindex="-1"
        >
           
          <item
            aria-selected="true"
            class="item is-selected"
            data-path="0"
            role="option"
          >
            <span>
              1
            </span>
             
          </item>
          <item
            aria-selected="false"
            class="item"
            data-path="1"
            role="option"
          >
            
             
            <p>
              2
            </p>
            
             
          </item>
          <item
            aria-selected="false"
            class="item"
            data-path="2"
            role="option"
          >
            
             
            <p>
              3
            </p>
            
             
          </item>
        </list>
      </scroll>
    </input-select>
  </div>
</body>
`;

exports[`Select.svelte > should render items using custom component 3`] = `
<body>
  <div>
    <input-select
      class="flex flex-col relative open"
      tabindex="0"
    >
      <selected-item
        aria-selected="false"
        class="w-full flex items-center"
        role="option"
        tabindex="-1"
      >
        <item>
          <span>
            1
          </span>
        </item>
         
        <iframe
          aria-hidden="true"
          src="about:blank"
          style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
          tabindex="-1"
        />
        <icon
          aria-label="opened"
          class="flex flex-shrink-0 items-center justify-center "
          role="button"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-opened"
          />
        </icon>
      </selected-item>
       
      <scroll
        class="flex flex-col absolute w-full z-10 overflow-scroll menu"
        style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both;"
      >
        <list
          class="flex flex-col w-full flex-shrink-0 select-none list"
          role="listbox"
          tabindex="-1"
        >
           
          <item
            aria-selected="true"
            class="item is-selected"
            data-path="0"
            role="option"
          >
            <span>
              1
            </span>
             
          </item>
          <item
            aria-selected="false"
            class="item"
            data-path="1"
            role="option"
          >
            <span>
              2
            </span>
             
          </item>
          <item
            aria-selected="false"
            class="item"
            data-path="2"
            role="option"
          >
            <span>
              3
            </span>
             
          </item>
        </list>
      </scroll>
    </input-select>
  </div>
</body>
`;

exports[`Select.svelte > should render items using custom mapping 1`] = `
<body>
  <div>
    <input-select
      class="flex flex-col relative open"
      tabindex="0"
    >
      <selected-item
        aria-selected="false"
        class="w-full flex items-center"
        role="option"
        tabindex="-1"
      >
        <item>
          
           
          <p>
            2
          </p>
          
        </item>
         
        <iframe
          aria-hidden="true"
          src="about:blank"
          style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
          tabindex="-1"
        />
        <icon
          aria-label="opened"
          class="flex flex-shrink-0 items-center justify-center "
          role="button"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-opened"
          />
        </icon>
      </selected-item>
       
      <scroll
        class="flex flex-col absolute w-full z-10 overflow-scroll menu"
        style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both;"
      >
        <list
          class="flex flex-col w-full flex-shrink-0 select-none list"
          role="listbox"
          tabindex="-1"
        >
           
          <item
            aria-selected="false"
            class="item"
            data-path="0"
            role="option"
          >
            
             
            <p>
              1
            </p>
            
             
          </item>
          <item
            aria-selected="true"
            class="item is-selected"
            data-path="1"
            role="option"
          >
            
             
            <p>
              2
            </p>
            
             
          </item>
          <item
            aria-selected="false"
            class="item"
            data-path="2"
            role="option"
          >
            
             
            <p>
              3
            </p>
            
             
          </item>
        </list>
      </scroll>
    </input-select>
  </div>
</body>
`;

exports[`Select.svelte > should render items using default component, when invalid component is provided 1`] = `
<body>
  <div>
    <input-select
      class="flex flex-col relative open"
      tabindex="0"
    >
      <selected-item
        aria-selected="false"
        class="w-full flex items-center"
        role="option"
        tabindex="-1"
      >
        <item>
          
           
          <p>
            2
          </p>
          
        </item>
         
        <iframe
          aria-hidden="true"
          src="about:blank"
          style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
          tabindex="-1"
        />
        <icon
          aria-label="opened"
          class="flex flex-shrink-0 items-center justify-center "
          role="button"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-opened"
          />
        </icon>
      </selected-item>
       
      <scroll
        class="flex flex-col absolute w-full z-10 overflow-scroll menu"
        style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both;"
      >
        <list
          class="flex flex-col w-full flex-shrink-0 select-none list"
          role="listbox"
          tabindex="-1"
        >
           
          <item
            aria-selected="false"
            class="item"
            data-path="0"
            role="option"
          >
            
             
            <p>
              1
            </p>
            
             
          </item>
          <item
            aria-selected="true"
            class="item is-selected"
            data-path="1"
            role="option"
          >
            
             
            <p>
              2
            </p>
            
             
          </item>
          <item
            aria-selected="false"
            class="item"
            data-path="2"
            role="option"
          >
            
             
            <p>
              3
            </p>
            
             
          </item>
        </list>
      </scroll>
    </input-select>
  </div>
</body>
`;

exports[`Select.svelte > should render items using field mapping 1`] = `
<body>
  <div>
    <input-select
      class="flex flex-col relative open"
      tabindex="0"
    >
      <selected-item
        aria-selected="false"
        class="w-full flex items-center"
        role="option"
        tabindex="-1"
      >
        <item>
           
          
        </item>
         
        <iframe
          aria-hidden="true"
          src="about:blank"
          style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
          tabindex="-1"
        />
        <icon
          aria-label="opened"
          class="flex flex-shrink-0 items-center justify-center "
          role="button"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-opened"
          />
        </icon>
      </selected-item>
       
      <scroll
        class="flex flex-col absolute w-full z-10 overflow-scroll menu"
        style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both;"
      >
        <list
          class="flex flex-col w-full flex-shrink-0 select-none list"
          role="listbox"
          tabindex="-1"
        >
           
          <item
            aria-selected="false"
            class="item"
            data-path="0"
            role="option"
          >
            
             
            <p>
              1
            </p>
            
             
          </item>
          <item
            aria-selected="false"
            class="item"
            data-path="1"
            role="option"
          >
            
             
            <p>
              2
            </p>
            
             
          </item>
          <item
            aria-selected="false"
            class="item"
            data-path="2"
            role="option"
          >
            
             
            <p>
              3
            </p>
            
             
          </item>
        </list>
      </scroll>
    </input-select>
  </div>
</body>
`;

exports[`Select.svelte > should render string array 1`] = `
<body>
  <div>
    <input-select
      class="flex flex-col relative open"
      tabindex="0"
    >
      <selected-item
        aria-selected="false"
        class="w-full flex items-center"
        role="option"
        tabindex="-1"
      >
        <item>
           
          
        </item>
         
        <iframe
          aria-hidden="true"
          src="about:blank"
          style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
          tabindex="-1"
        />
        <icon
          aria-label="opened"
          class="flex flex-shrink-0 items-center justify-center "
          role="button"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-opened"
          />
        </icon>
      </selected-item>
       
      <scroll
        class="flex flex-col absolute w-full z-10 overflow-scroll menu"
        style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both;"
      >
        <list
          class="flex flex-col w-full flex-shrink-0 select-none list"
          role="listbox"
          tabindex="-1"
        >
           
          <item
            aria-selected="false"
            class="item"
            data-path="0"
            role="option"
          >
             
            <p>
              a
            </p>
            
             
          </item>
          <item
            aria-selected="false"
            class="item"
            data-path="1"
            role="option"
          >
             
            <p>
              b
            </p>
            
             
          </item>
          <item
            aria-selected="false"
            class="item"
            data-path="2"
            role="option"
          >
             
            <p>
              c
            </p>
            
             
          </item>
        </list>
      </scroll>
    </input-select>
  </div>
</body>
`;

exports[`Select.svelte > should render string array 2`] = `
<body>
  <div>
    <input-select
      class="flex flex-col relative"
      tabindex="0"
    >
      <selected-item
        aria-selected="true"
        class="w-full flex items-center"
        role="option"
        tabindex="-1"
      >
        <item>
           
          <p>
            b
          </p>
          
        </item>
         
        <iframe
          aria-hidden="true"
          src="about:blank"
          style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
          tabindex="-1"
        />
        <icon
          aria-label="closed"
          class="flex flex-shrink-0 items-center justify-center "
          role="button"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-closed"
          />
        </icon>
      </selected-item>
       
      <scroll
        class="flex flex-col absolute w-full z-10 overflow-scroll menu"
        style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both, __svelte_2150010166_0 50ms linear 0ms 1 both;"
      >
        <list
          class="flex flex-col w-full flex-shrink-0 select-none list"
          role="listbox"
          tabindex="-1"
        >
           
          <item
            aria-selected="false"
            class="item"
            data-path="0"
            role="option"
          >
             
            <p>
              a
            </p>
            
             
          </item>
          <item
            aria-selected="false"
            class="item is-selected"
            data-path="1"
            role="option"
          >
             
            <p>
              b
            </p>
            
             
          </item>
          <item
            aria-selected="false"
            class="item"
            data-path="2"
            role="option"
          >
             
            <p>
              c
            </p>
            
             
          </item>
        </list>
      </scroll>
    </input-select>
  </div>
</body>
`;
